slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java

```diff
@@ -61,4 +66,45 @@ public static boolean areEquivalent(Tree first, Tree second) {
 
     return true;
   }
+
+  public static List<List<Tree>> findDuplicatedGroups(List<Tree> list) {
+    return list.stream()
+      .collect(Collectors.groupingBy(ComparableTree::new, LinkedHashMap::new, Collectors.toList()))
+      .values().stream()
+      .filter(group -> group.size() > 1)
+      .collect(Collectors.toList());
+  }
+
+  static class ComparableTree {
+
+    private final Tree tree;
+    private final int hash;
+
+    ComparableTree(Tree tree) {
+      this.tree = tree;
+      hash = computeHash(tree);
+    }
+
+    private static int computeHash(@Nullable Tree tree) {
+      if (tree == null) {
+        return 0;
+      }
+      return TreePrinter.tree2string(tree).hashCode();
```

Comment:
`tree2string` method does not implement printing for new trees (`BlockTree`, `FunctionDeclarationTree`, `IfTree`, `MatchTree` and `MatchCaseTree`), so we may have various surprises there...

Reply:
I think it's ok.